---
title: "Amplicon Data Analysis"
author: "Andreas Schneider"
date: "30/08/2018"
output:
  html_document:
    number_sections: true
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#ITS dataset

## Normalization of count matrix with metagenomeSeq

The count matrix and taxonomy table (with all OTUs) are derived from the ouput of the pre-processing pipeline and the OTU clustering scripts. OTUs
needed to have 5 reads in at least 2 biological replicates to be kept for the following analyses.

```{r include = FALSE}
library(metagenomeSeq)
library(phyloseq)
library(ggplot2)
```


```{r}

Seq_meta = loadPhenoData("../../doc/meta.csv", tran = TRUE)
Seq_meta = Seq_meta[as.integer(Seq_meta$Reads_Remaining)>4000,]
Seq_counts = loadMeta("../../doc/countmatrix.csv")
Seq_counts$counts = Seq_counts$counts[,colnames(Seq_counts$counts) %in% rownames(Seq_meta)]
Seq_taxa = read.delim("../../doc/tax.csv", stringsAsFactors = FALSE)
ord = match(colnames(Seq_counts$counts), rownames(Seq_meta))
Seq_meta = Seq_meta[ord, ]
ann_taxa = AnnotatedDataFrame(Seq_taxa)
ann_meta = AnnotatedDataFrame(Seq_meta)
EXP = newMRexperiment(Seq_counts$counts, phenoData = ann_meta, featureData = ann_taxa)
p=cumNormStat(EXP)
css_mat = cumNormMat(EXP,p=p)
rownames(css_mat) = Seq_taxa$SHA1

##summarize to Class level
###Exclude unknown and useless annotations
known_class = which(fData(EXP)$Class!="unidentified"
                    &fData(EXP)$Class!="Unknown"
                    &fData(EXP)$Class!="Incertae sedis"
                    &fData(EXP)$Class!="Insecta"
                    &fData(EXP)$Class!="GS17"
                    &fData(EXP)$Class!="GS18"
                    &fData(EXP)$Class!="Mortierellomycotina_cls_Incertae_sedis")
EXP_known_c = EXP[known_class,]
EXP_class = aggTax(EXP_known_c, lvl="Class")
p_class = cumNormStat(EXP_class)
mat_class = aggTax(EXP_known_c,lvl="Class",norm=TRUE,out="matrix")


#set colors that will be used throughout analysis
cols = c("#0099ff",
         "#c364c5",
         "#b9a633",
         "#4c4c4c",
         "#1cd3a2",
         "#fc2847",
         "#7e904a",
         "#5d76cb",
         "#ff8243",
         "#4ab0cc",
         "#d14448",
         "#dfe100",
         "#bf7854")
names(cols) = sort(unique(as.character(Seq_meta$Description_ID)))

```

## Data Analysis

### Most abundant species in ITS dataset
This result was used in  **Table S4**

```{r}
spru_vec = which(Seq_meta$Species=="Spruce")
asp_vec = which(Seq_meta$Species=="Aspen")
most_abundant_spru = data.frame(rowSums(css_mat[,spru_vec]))
most_abundant_spru[,2] = rownames(most_abundant_spru)
colnames(most_abundant_spru) = c("Norm_Counts_Spruce", "SHA1")
most_abundant_spru = merge.data.frame(most_abundant_spru, Seq_taxa[,c(2:9)])
most_abundant_spru = most_abundant_spru[order(most_abundant_spru$Norm_Counts_Spruce, decreasing = TRUE),]
most_abundant_spru_f = most_abundant_spru[most_abundant_spru$Genus!="unidentified",]
most_abundant_spru_f[1:25,]
```

### PCoA of Bray-Curtis dissimilarities between samples

These results were used in **Fig 3** and **Fig S3**

The package phyloseq is used to plot the PCoA

```{r}
taxonomy3 = Seq_taxa
taxonomy3$OTU = as.character(taxonomy3$OTU)
rownames(taxonomy3) = Seq_taxa$SHA1 
taxonomy3 = as.matrix(taxonomy3)
rownames(css_mat) = rownames(taxonomy3)


ps = phyloseq(tax_table(taxonomy3), 
              otu_table(css_mat,taxa_are_rows = TRUE),
              sample_data(Seq_meta))

#cleaning
Seq_meta$tissue.type[Seq_meta$Description=="Friggesund seedlings"] = "seedlings"
Seq_meta$tissue.type = as.factor(Seq_meta$tissue.type)
#Seq_meta$ID_files = rownames(Seq_meta)

#subset phyloseq object to spruce only
ps_spru = prune_samples(Seq_meta$Species=="Spruce", ps)

#ordinate
out.bc.log_spru = ordinate(ps_spru, method = "MDS", distance = "bray")

#subset to aspen
ps_asp = prune_samples(Seq_meta$Species=="Aspen", ps)

#ordinate
out.bc.log_asp = ordinate(ps_asp, method = "MDS", distance = "bray")

##define plotting function for PCoA plots
plot_pcoa = function (ps, out, cols) {
  cols_s = cols[sort(unique(as.character(sample_data(ps)$Description_ID)))]
  plot_ordination(ps, out, color = "Description_ID", shape = "tissue.type")+
    geom_point(size = 5, alpha = 0.7)+
    labs(col = "Samples")+
    scale_color_manual(values = cols_s)+
    theme_classic()+
    theme(axis.text = element_text(face = "bold", size = 15),
        axis.line = element_line(size = 1, linetype = "solid"),
        axis.title = element_text(size = 15, face = "bold"),
        #axis.ticks = element_line(size = 1, linetype = "solid"),
        legend.text = element_text(size=15, face = "bold"),
        legend.title = element_text(size = 17, face = "bold"),
        plot.title = element_text(size = 21, face = "bold"),
        strip.text = element_text(size = 19, face = "bold"))#,
        #legend.position = "none")
}

#Plot spruce samples
plot_pcoa(ps_spru, out.bc.log_spru, cols)

#Plot aspen samples
plot_pcoa(ps_asp, out.bc.log_asp, cols)


```

Spruce ITS samples, counts aggregated by class, PcoA ordination (**Fig S4**)

```{r}
ps_c = phyloseq(otu_table(mat_class[,Seq_meta$Species=="Spruce"], taxa_are_rows = TRUE),
                sample_data(Seq_meta[Seq_meta$Species=="Spruce",]))

pcoa_c = ordinate(ps_c, method = "MDS", distance = "bray")

plot_pcoa(ps_c, pcoa_c)
```

# 16S dataset

## Normalization of count matrix with metagenomeSeq

Description: see above.

```{r}

Seq_16S_meta = loadPhenoData("../../doc/16Smeta.csv", tran = TRUE)
Seq_16S_meta = Seq_16S_meta[as.integer(Seq_16S_meta$Reads_Remaining)>4000,]
Seq_16S_counts = loadMeta("../../doc/16Scountmatrix.csv")
Seq_16S_counts$counts = Seq_16S_counts$counts[,colnames(Seq_16S_counts$counts) %in% rownames(Seq_16S_meta)]
Seq_16S_taxa = read.delim("../../doc/16Stax.csv", stringsAsFactors = FALSE)
ord_16S = match(colnames(Seq_16S_counts$counts), rownames(Seq_16S_meta))
Seq_16S_meta = Seq_16S_meta[ord_16S, ]
ann_16S_taxa = AnnotatedDataFrame(Seq_16S_taxa)
ann_16S_meta = AnnotatedDataFrame(Seq_16S_meta)
EXP_16S = newMRexperiment(Seq_16S_counts$counts, phenoData = ann_16S_meta, featureData = ann_16S_taxa)
p_16S=cumNormStat(EXP_16S)
css_mat_16S = cumNormMat(EXP_16S,p=p_16S)
rownames(css_mat_16S) = Seq_16S_taxa$SHA1

##summarize to Class level
###Exclude unknown and useless annotations
known_class_16S = which(fData(EXP_16S)$Class!="uncultured bacterium"
                    &fData(EXP_16S)$Class!="Unknown"
                    &fData(EXP_16S)$Class!="ARKICE-90"
                    &fData(EXP_16S)$Class!="OPB35 soil group"
                    &fData(EXP_16S)$Class!="Ambiguous_taxa"
                    &fData(EXP_16S)$Class!="Bacteroidetes Incertae Sedis"
                    &fData(EXP_16S)$Class!="vadinHA49"
                    &fData(EXP_16S)$Class!="Verrumicrobia Incertae Sedis")
EXP_16S_known_c = EXP_16S[known_class_16S,]
EXP_16S_class = aggTax(EXP_16S_known_c, lvl="Class")
p_class_16S = cumNormStat(EXP_16S_class)
mat_class_16S = aggTax(EXP_16S_known_c,lvl="Class",norm=TRUE,out="matrix")

meta_col = read.csv("../../doc/meta_COL.csv")
rownames(meta_col) = meta_col$Sample

cols2 = as.character(meta_col$Colour)
cols2 = unique(cols2)
names(cols2) = as.character(unique(meta_col$Description_ID))

```

## Data Analysis

### PCoA of Bray-Curtis dissimilarities between samples

```{r}
taxonomy4 = Seq_16S_taxa
taxonomy4$OTU = as.character(taxonomy4$OTU)
rownames(taxonomy4) = Seq_16S_taxa$SHA1 
taxonomy4 = as.matrix(taxonomy4)
rownames(css_mat_16S) = rownames(taxonomy4)

ps_16S = phyloseq(tax_table(taxonomy4), 
              otu_table(css_mat_16S,taxa_are_rows = TRUE),
              sample_data(Seq_16S_meta))

#cleaning
Seq_16S_meta$tissue.type[Seq_16S_meta$Description=="Friggesund seedlings"] = "seedlings"
Seq_16S_meta$tissue.type = as.factor(Seq_16S_meta$tissue.type)

#subset phyloseq object to spruce only
Seq_16S_meta$Species = ifelse(grepl("^A", Seq_16S_meta$Sort_ID), "Aspen", "Spruce")
ps_16S_spru = prune_samples(Seq_16S_meta$Species=="Spruce", ps_16S)

#ordinate
out.bc.log_16S_spru = ordinate(ps_16S_spru, method = "MDS", distance = "bray")

#subset to aspen
ps_16S_asp = prune_samples(Seq_16S_meta$Species=="Aspen", ps_16S)

#ordinate
out.bc.log_16S_asp = ordinate(ps_16S_asp, method = "MDS", distance = "bray")


#Plot spruce samples
plot_pcoa(ps_16S_spru, out.bc.log_16S_spru, cols2)

#Plot aspen samples
plot_pcoa(ps_16S_asp, out.bc.log_16S_asp, cols2)

```

