# Reproducibility and biological information of tree-associated metacommunities using RNA Sequencing and DNA amplicon profiling
## Authors
Görel Sundström§1, Andreas N Schneider§2, Kerstin Richau2, Julia C Haas2, Nicolas Delhomme1, Andreas Sjödin3,4, Vaughan Hurry1, Manfred Grabherr5*, Nathaniel R Street2&ast;

§ Equal contribution 

&ast; corresponding author

## Affiliations

1 Umeå Plant Science Centre, Department of Forest Genetics and Plant Physiology, SLU, SE-901 85, Umeå, Sweden 

2 Umeå Plant Science Centre, Department of Plant Physiology, Umeå University, SE-901 87, Umeå, Sweden 

3 Division of CBRN Security and Defence, FOI–Swedish Defence Research Agency, SE-906 21, Umeå, Sweden  

4 Department of Chemistry, Computational Life Science Cluster (CLiC), Umeå University, SE-901 87, Umeå, Sweden 

5 Department of Medical Biochemistry and Microbiology, Uppsala University, Uppsala 

## Abstract

**Background:** Health, growth and fitness of boreal forest trees is impacted and improved by their associated fungal and bacterial microbiomes. Microbial, as well as plant host gene expression and functional activity can be assayed with RNA sequencing (RNA-Seq) data from host samples. In contrast, phylogenetic marker gene amplicon sequencing data is used to assess taxonomic composition and community structure of the microbiome. However, it is not clear how much of this structural and taxonomic information is included in transcriptomic data from the same sample. Here we investigate how comparable the conclusions are that can be drawn about relative community structure and taxonomic composition using host-derived RNA-Seq and DNA amplicon (bacterial 16S and fungal ITS1) sequencing data. As a study system we used a panel of phyllospheric samples from two Swedish boreal forest tree species, Picea abies (Norway spruce) and Populus tremula (Eurasian aspen). 

**Results:** Fungal and bacterial transcriptomes showed correlating differences among tree tissues. Fungal gene expression correlated with community structure of the ITS data, but bacterial gene expression counts did not correlate with 16S data. Direct comparison of taxonomic units (classes) between transcriptomic and amplicon data revealed many common classes, but also classes only detected by one of the methods, which is likely due to contrasting and insufficient database representation. RNA-Seq data can give additional insights into active biological pathways in each sample. 

**Conclusions:** In the future, increasing database coverage will enable RNA-Seq data to also answer questions about community structure and taxonomic composition of fungal communities, but cannot completely replace amplicon sequencing data. 

 

*Keywords:* Norway spruce, Eurasian aspen, Microbiome, Phyllosphere, metatranscriptomics, RNA sequencing, amplicon sequencing, 16S rRNA gene 

## Content of this repository

This repository contains the custom scripts and associated metadata mentioned in the manuscript methods. 
